public/js/src/module/offline-app-worker.js

```diff
@@ -0,0 +1,137 @@
+const STATIC_CACHE = 'enketo-common';
+const FORMS_CACHE = 'enketo-forms';
+
+/**
+ * @param {string} url
+ */
+const cacheName = (url) => {
+    if (
+        url === '/favicon.ico' ||
+        url.endsWith('/x/') ||
+        /\/x\/((css|fonts|images|js|locales)\/|offline-app-worker.js)/.test(url)
+    ) {
+        return STATIC_CACHE;
+    }
+
+    return FORMS_CACHE;
+};
+
+/**
+ * @param {Request | string} key
+ * @param {Response} response
+ */
+const cacheResponse = async (key, response) => {
+    const clone = response.clone();
+    const cache = await caches.open(cacheName(key.url ?? key));
+
+    await cache.put(key, clone);
+
+    return response;
+};
+
+/**
+ * @param {Response} response
+ */
+const cachePrefetchURLs = async (response) => {
+    const linkHeader = response.headers.get('link') ?? '';
+    const prefetchURLs = [
+        ...linkHeader.matchAll(/<([^>]+)>;\s*rel="prefetch"/g),
+    ].map(([, url]) => url);
+
+    const cache = await caches.open(STATIC_CACHE);
+
+    await Promise.allSettled(prefetchURLs.map((url) => cache.add(url)));
+};
+
+self.addEventListener('install', () => {
+    self.skipWaiting();
+});
+
+const removeStaleCaches = async () => {
+    const cacheStorageKeys = await caches.keys();
+
+    cacheStorageKeys.forEach((key) => {
+        if (key !== FORMS_CACHE) {
+            caches.delete(key);
+        }
+    });
+};
+
+const onActivate = async () => {
+    await self.clients.claim();
+    await removeStaleCaches();
+};
+
+self.addEventListener('activate', (event) => {
+    event.waitUntil(onActivate());
+});
+
+const FETCH_OPTIONS = {
+    cache: 'reload',
+    credentials: 'same-origin',
+};
+
+/**
+ * @param {Request} request
+ */
+const onFetch = async (request) => {
+    const { method, referrer, url } = request;
+
+    if (method !== 'GET') {
+        return fetch(request, FETCH_OPTIONS);
+    }
+
+    const isFormPageRequest =
+        url.includes('/x/') && (referrer === '' || referrer === url);
+    const cacheKey = isFormPageRequest ? url.replace(/\/x\/.*/, '/x/') : url;
```

Comment:
I think this concept of caching the HTML under a single key merits some kind of comment.